src/attr/_make.py

```diff
@@ -483,6 +513,13 @@ def __init__(
             self._cls_dict["__setattr__"] = _frozen_setattrs
             self._cls_dict["__delattr__"] = _frozen_delattrs
 
+        if (
+            cache_hash
+            and _method_eq(cls.__reduce__, object.__reduce__)
```

Comment:
Note to reviewers: This is probably the most controversial part - the `__reduce__` is generated only if you haven't defined your own `__reduce__` or `__reduce_ex__`.

I have done this because it seems mildly inappropriate to squash someone's existing `__reduce__` and I am not confident that we can count on the third return value from a *custom* `__reduce__` being the object's `__dict__`.